lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,272 @@
+      this._requestBodyLength += data.length;
+      this._requestBodyBuffers.push({ data, encoding });
+      this._currentRequest.write(data, encoding);
+    }
+  }
+
+  end() {
+    this.emit("request", this._currentRequest);
+    this._ended = this._ending = true;
+    this._currentRequest.end();
+  }
+
+  setHeader(name, value) {
+    this._options.headers[name] = value;
+    this._currentRequest.setHeader(name, value);
+  }
+
+  removeHeader(name) {
+    delete this._options.headers[name];
+    this._currentRequest.removeHeader(name);
+  }
+
+  toJSON() {
+    const uri = new URL(this._currentUrl);
+    const { method, headers } = this._options;
+    return { uri, method, headers };
+  }
+
+  _performRequest() {
+    const urlOptions = new URL(this._currentUrl);
+    const { protocol: scheme } = urlOptions;
+    this._options.agent = this._options.agents[scheme.substr(0, scheme.length - 1)];
+    const self = this;
+    function onNative(response) {
+      self._processResponse(response);
+    }
+    const { request } = scheme === "https:" ? https : http;
+    const message = request(this._currentUrl, this._options, onNative);
+    this._currentRequest = message;
+    this.getCookie();
+    ["connect", "error", "socket", "timeout"].forEach(event => {
+      message.on(event, (...args) => {
+        self.emit(event, ...args);
+      });
+    });
+    if (this._isRedirect) {
+      let i = 0;
+      const buffers = this._requestBodyBuffers;
+      (function writeNext(error) {
+        if (message === self._currentRequest) {
+          if (error) {
+            self.emit("error", error);
+          } else if (i < buffers.length) {
+            const buffer = buffers[i++];
+            if (!message.writableEnded) {
+              message.write(buffer.data, buffer.encoding, writeNext);
+            }
+          } else if (self._ended) {
+            message.end();
+          }
+        }
+      })();
+    }
+  }
+
+  _processResponse(response) {
+    response.request = {};
+    response.request.headers = this._options.headers;
+    response.request.uri = new URL(this._currentUrl);
+    this.response = response;
+    this.setCookies(response);
+    const { statusCode } = response;
+    const { location } = response.headers;
+    const catchResErrors = err => {
+      if (!(majorNodeVersion === 15 && err.message === "aborted")) {
+        this.emit("error", err);
+      }
+    };
+    response.on("error", catchResErrors);
+    let redirectAddress = null;
+    if (typeof location === "string" &&
+      location.length &&
+      this._options.followRedirects &&
+      statusCode >= 300 &&
+      statusCode < 400) {
+      redirectAddress = location;
+    } else if (statusCode === 401 &&
+      /^Basic /i.test(response.headers["www-authenticate"] || "") &&
+      (this._options.user && this._options.user.length)) {
+      const { user, pass } = this._options;
+      this._options.auth = `${user}:${pass}`;
+      redirectAddress = 1;
```

Comment:
1??? It seems like this variable is serving two purposes and you should instead use two different variables.